File: bindings/python/examples/dht.py
#!/usr/bin/env python3
import sys
import time
import random
from lattica import Lattica

def main():
    args = sys.argv[1:]
    bootstrap_nodes = args if args else []
    print(f"Bootstrap nodes: {bootstrap_nodes}")

    if bootstrap_nodes:
        lattica_inst = Lattica.builder() \
            .with_bootstraps(bootstrap_nodes) \
            .build()

        try:
            print("\n=== Basic DHT ===")
            lattica_inst.store("name", "ween") # default expiration
            result = lattica_inst.get("name")
            if result:
                print(f"Retrieved value: {result.value}, expiration: {result.expiration_time}")
            else:
                print("No record found")

            print("\n=== Subkey DHT ===")
            key = "peer_list"
            peers = ["alice", "bob", "carol", "david"]

            expiration_time = time.time() + 600
            for peer in peers:
                vote = random.choice(["yes", "no", "maybe"])
                lattica_inst.store(key, vote, expiration_time, subkey=peer)

            # wait for validity
            time.sleep(2)
            votes_result = lattica_inst.get(key)

            if votes_result:
                print("All votes:")
                for peer, vote_info in votes_result.value.items():
                    print(f"  {peer}: {vote_info.value} expires: {vote_info.expiration_time}")
            else:
                print("No votes found")

        except Exception as e:
            print(f"Error during DHT operations: {e}")
            import traceback
            traceback.print_exc()
    else:
        print("\n=== Running as DHT Server ===")
        print("Waiting for connections... (Press Ctrl+C to stop)")
        Lattica.builder().build()
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            print("Server shutting down...")

# node1: python dht.py
# node2: python dht.py /ip4/127.0.0.1/tcp/x/p2p/x
if __name__ == "__main__":
    main()

File: lattica/examples/dht.rs
use lattica::{network, common};
use std::time::{Duration, Instant};
use libp2p::kad::{Quorum, Record, RecordKey};
use tokio::signal;
use std::env;
use libp2p::Multiaddr;

// node1: cargo run --example dht
// node2: cargo run --example dht /ip4/127.0.0.1/tcp/x/p2p/xxxxxxxx
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt().with_max_level(tracing::Level::INFO).init();
    
    let args: Vec<String> = env::args().collect();
    let bootstrap_nodes = if args.len() > 1 {
        args[1..].into_iter().filter_map(|addr| {
            match addr.parse::<Multiaddr>() {
                Ok(addr) => Some(addr),
                Err(e) => {
                    tracing::warn!("Invalid bootstrap address '{}': {}", addr, e);
                    None
                }
            }
        }).collect()
    } else {
        vec![]
    };

    let mut lattica = network::Lattica::builder()
        .with_bootstrap_nodes(bootstrap_nodes.clone())
        .build().await?;

    tokio::time::sleep(Duration::from_millis(1000)).await;

    // client node
    if !bootstrap_nodes.is_empty() {
        let current_time = common::time::get_current_timestamp();
        let expiration_time = current_time + 600.0;
        tracing::info!("1. store simple");
        lattica.store_simple("greeting", "Hello, World!".as_bytes().to_vec(), expiration_time).await?;
        if let Some(value) = lattica.get_with_subkey("greeting").await? {
            match value {
                common::types::DhtValue::Simple { value, expiration } => {
                    let content = String::from_utf8_lossy(&value);
                    tracing::info!("get simple value: '{}', exp: {}", content, expiration);
                }
                _ => {}
            }
        }

        tracing::info!("2. store with sub key");
        lattica.store_subkey("party_vote", "alice", "yes".as_bytes().to_vec(), expiration_time).await?;
        lattica.store_subkey("party_vote", "bob", "yes".as_bytes().to_vec(), expiration_time).await?;
        lattica.store_subkey("party_vote", "carol", "no".as_bytes().to_vec(), expiration_time).await?;
        lattica.store_subkey("party_vote", "david", "maybe".as_bytes().to_vec(), expiration_time).await?;
        tokio::time::sleep(Duration::from_millis(1000)).await;

        if let Some(value) = lattica.get_with_subkey("party_vote").await? {
            match value {
                common::types::DhtValue::WithSubkeys { subkeys } => {
                    tracing::info!("get with subkeys:");
                    for (subkey, vote_bytes, expiration) in subkeys {
                        let vote = String::from_utf8_lossy(&vote_bytes);
                        tracing::info!("  {}: {} (expiration: {})", subkey, vote, expiration);
                    }
                }
                _ => {}
            }
        } else {
            tracing::info!("get with no subkeys");
        }

        tracing::info!("3. store complex data");
        let user_status = serde_json::json!({
            "online": true,
            "last_seen": current_time,
            "mood": "happy"
        });
        let user_profile = serde_json::json!({
            "name": "Alice",
            "age": 25,
            "city": "Beijing"
        });
        lattica.store_subkey("user_status", "user_123", user_status.to_string().as_bytes().to_vec(), expiration_time).await?;
        lattica.store_subkey("user_status", "user_456", user_profile.to_string().as_bytes().to_vec(), expiration_time).await?;
        tokio::time::sleep(Duration::from_millis(100)).await;
        if let Some(value) = lattica.get_with_subkey("user_status").await? {
            match value {
                common::types::DhtValue::WithSubkeys { subkeys } => {
                    tracing::info!("user status:");
                    for (user_id, data_bytes, expiration) in subkeys {
                        let data_str = String::from_utf8_lossy(&data_bytes);
                        if let Ok(json_data) = serde_json::from_str::<serde_json::Value>(&data_str) {
                            tracing::info!("  {}: {} (expiration: {})", user_id, json_data, expiration);
                        } else {
                            tracing::info!("  {}: {} (expiration: {})", user_id, data_str, expiration);
                        }
                    }
                }
                _ => {}
            }
        }

        tracing::info!("4. key expired");
        let short_expiration = current_time + 2.0;
        lattica.store_simple("temp_data", "expired after 2 seconds".as_bytes().to_vec(), short_expiration).await?;

        tracing::info!("store，expired after 2 seconds...");
        tokio::time::sleep(Duration::from_secs(3)).await;

        match lattica.get_with_subkey("temp_data").await? {
            Some(_) => tracing::info!("expired failed"),
            None => tracing::info!("expired success"),
        }
    }

    signal::ctrl_c().await.expect("failed to listen for event");
    Ok(())
}